qemu/deps/performance/start_testpmd.py

```diff
@@ -0,0 +1,155 @@
+            return False
+
+    def start(self):
+        self.command("start")
+
+    def stop(self):
+        self.command("stop")
+
+    def quit(self):
+        self.proc.sendline("quit")
+        logging.info("testpmd> quit")
+        print("testpmd> quit")
+        self.proc.expect("Bye...")
+        #self.proc.expect([pexpect.TIMEOUT, 'Bye...', pexpect.EOF])
+        logging.info(self.proc.before)
+        line_list = to_text(self.proc.before).split("\n")
+        for subline in line_list:
+            if len(subline.strip()) > 0:
+                print(subline)
+        return to_text(self.proc.before)
+
+    def set_port_stats(self):
+        self.command("show port stats all")
+
+    def set_portlist(self, portlist):
+        self.command("set portlist %s" % portlist)
+
+    def get_config_fwd(self):
+        self.command("show config fwd")
+
+    def set_fwd_mac_retry(self):
+        self.command('set fwd mac retry')
+
+    def set_vlan_0(self):
+        self.command('vlan set strip on 0')
+
+    def set_vlan_1(self):
+        self.command('vlan set strip on 1')
+
+    def command(self, cmd):
+        self.proc.sendline(cmd)
+        #self.proc.expect([pexpect.TIMEOUT, 'testpmd>', pexpect.EOF])
+        self.proc.expect("testpmd>")
+        logging.info("testpmd> %s" % cmd)
+        print("testpmd> %s" % cmd)
+        logging.info(self.proc.before)
+        line_list = to_text(self.proc.before).split("\n")
+        for subline in line_list:
+            if len(subline.strip()) > 0:
+                print(subline)
+
+        return to_text(self.proc.before)
+
+
+def start_testpmd(nic1_driver, nic2_driver, nic1, nic2, cores, queues):
+
+    my_testpmd = TestPMD()
+    my_testpmd.launch(
+        nic1_driver=nic1_driver,
+        nic2_driver=nic2_driver,
+        nic1=nic1,
+        nic2=nic2,
+        cores=cores,
+        queues=queues)
+
+    my_testpmd.set_fwd_mac_retry()
+    my_testpmd.set_vlan_0()
+    my_testpmd.set_vlan_1()
+
+    my_testpmd.start()
+    my_testpmd.set_port_stats()
+
+    # testmpd will quit after running_time
+    start_time = time.time()
+    end_time = start_time + running_time
+    while time.time() < end_time:
+        time.sleep(10)
+        print("time.time=%s" % time.time)
+    my_testpmd.stop()
+    my_testpmd.set_port_stats()
+    my_testpmd.quit()
+
+
+def to_text(data, encoding=ENCODING):
+    """
+    Convert anything to text decoded text
+
+    When the data is bytes, it's decoded. When it's not of string types
+    it's re-formatted into text and returned. Otherwise (it's string)
+    it's returned unchanged.
+
+    :param data: data to be transformed into text
+    :param encoding: encoding of the data (only used when decoding
+                     is necessary)
+    :type data: either bytes or other data that will be returned
+                unchanged
+    """
+    if isinstance(data, bytes):
+        if encoding is None:
+            encoding = ENCODING
+        return data.decode(ENCODING)
```

Comment:
Here should be `data.decode(encoding)`? Otherwise why you define `encoding`?

Reply:
If `return data.decode(ENCODING)`, then Line 143-145 is useless so should be removed.